cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cube/DefaultCube.java

```diff
@@ -113,8 +122,18 @@ public void add(Collection<? extends CubeFact> facts) {
       }
     }
 
-    for (FactTable table : resolutionToFactTable.values()) {
-      table.add(toWrite);
+    Map<Integer, Future> futures = new HashMap<>();
```

Comment:
Seems like you are not using the key in this map at all. Why not just a list?

Reply:
This is to identify which resolution table fails to add, https://github.com/cdapio/cdap/blob/36830b687f400f4aa25bb4eee47b1b1a2d86a5bc/cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cube/DefaultCube.java#L135